utils/forest_utils/src/miscs/either.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::AsyncRead;
+
+pub enum Either<L, R> {
+    Left(L),
+    Right(R),
+}
+
+impl<L, R> Either<L, R> {
+    fn left_mut(&mut self) -> Option<&mut L> {
+        match self {
+            Self::Left(left) => Some(left),
+            _ => None,
+        }
+    }
+
+    fn right_mut(&mut self) -> Option<&mut R> {
+        match self {
+            Self::Right(right) => Some(right),
+            _ => None,
+        }
+    }
+}
+
+impl<L: AsyncRead + Unpin, R: AsyncRead + Unpin> AsyncRead for Either<L, R> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::io::Result<usize>> {
+        match self.deref_mut() {
+            Self::Left(left) => Pin::new(left).poll_read(cx, buf),
+            Self::Right(right) => Pin::new(right).poll_read(cx, buf),
+        }
```

Comment:
Is this code correct? Can you freely create new pins like this?

Reply:
Don't you need [map_unchecked_mut](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.map_unchecked_mut) or something? I'm not an expert on this.

Reply:
It does compile, so I guess it works. Will c if sync check CI job passes

Reply:
@LesnyRumcajs @elmattic What are your thoughts on this? I feel this documentation is relevant: https://doc.rust-lang.org/std/pin/index.html#projections-and-structural-pinning

Reply:
Oh, I see. You require both `L` and `R` to be `Unpin`.

Reply:
`map_unchecked_mut` definitely looks like a better fit but it's `unsafe`, is it allowed to use `unsafe` in `forest`?

Reply:
You probably want to use `Pin::into_inner` instead of `deref_mut`, then.

Reply:
You can use unsafe functions as long as you prove the correctness yourself. :)